File: ctci/chapter_8_recursion_dp/python/8.3-magic_index.py
from typing import List


def mfh(arr: List, start: int, end: int) -> int:
    """
    Helper.
    """
    if start > end:
        return -1
    mid_index = (start + end) // 2
    mid_val = arr[mid_index]
    if mid_index == mid_val:
        return mid_index
    # Search left
    left_index = min(mid_index - 1, mid_val)
    left = mfh(arr, 0, left_index)
    if left >= 0:
        return left
    # Search right
    right_index = max(mid_index + 1, mid_val)
    right = mfh(arr, right_index, end)
    return right


def magic_fast(arr: List) -> int:
    """
    Finds the magic index where index == val at index. Values do NOT have to
    be distinct.
    """
    return mfh(arr, 0, len(arr) - 1)


arr = [-10, -5, 2, 2, 2, 3, 4, 7, 9, 12, 13]
print(magic_fast(arr))  # 2

arr = [-10, -5, -1, 0, 2, 3, 4, 7, 9, 12, 13]
print(magic_fast(arr))  # 7

arr = [-10, -5, -1, 0, 2, 3, 4, 8, 9, 12, 13]
print(magic_fast(arr))  # -1

File: ctci/tmp/chapter_8_recursion_dp/js/8.3-magicIndex.js
/**
 * Finds the magic index where index == val at index. Values do NOT have to
 * be distinct.
 * @param   {number[]}  arr
 * @returns {number}
 */
const magic_fast = arr => {
  return mfh(arr, 0, arr.length - 1)
}

/**
 * Helper func
 * @param   {number[]}  arr
 * @param   {number}    start
 * @param   {number}    end
 * @returns {number}
 */
const mfh = (arr, start, end) => {
  if (start > end) return -1
  const mid_index = Math.floor((start + end) / 2)
  const mid_val = arr[mid_index]
  if (mid_val === mid_index) return mid_index
  /* Search left */
  const left_index = Math.min(mid_index - 1, mid_val)
  const left = mfh(arr, start, left_index)
  if (left >= 0) return left
  /* Search right */
  const right_index = Math.max(mid_index + 1, mid_val)
  const right = mfh(arr, right_index, end)
  return right
}

let arr = [-10, -5, 2, 2, 2, 3, 4, 7, 9, 12, 13]
console.log(magic_fast(arr)) // 2

arr = [-10, -5, -1, 0, 2, 3, 4, 7, 9, 12, 13]
console.log(magic_fast(arr)) // 7

arr = [-10, -5, -1, 0, 2, 3, 4, 8, 9, 12, 13]
console.log(magic_fast(arr)) // -1
